runtime/src/gpu/nvidia/gpu-nvidia.c

```diff
@@ -352,6 +352,14 @@ void* chpl_gpu_impl_mem_alloc(size_t size) {
 void chpl_gpu_impl_mem_free(void* memAlloc) {
   if (memAlloc != NULL) {
     assert(chpl_gpu_is_device_ptr(memAlloc));
+
+    // see note in chpl_gpu_mem_free
+    int64_t dev_id = c_sublocid_none;
+    CUDA_CALL(cuPointerGetAttribute((void*)&dev_id,
+                                    CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL,
+                                    (CUdeviceptr)memAlloc));
+    switch_context(dev_id);
```

Comment:
I'm assuming that when you switch context to a context that is already active this is a no-op right?

Reply:
I don't have a general answer. For NVIDIA, we make sure to not do anything ourselves. So, there's a check involved which could count as an operation: https://github.com/chapel-lang/chapel/blob/main/runtime/src/gpu/nvidia/gpu-nvidia.c#L85-L89

For AMD: we simply set the device as we don't do much with contexts on the AMD side (for the better, I think): https://github.com/chapel-lang/chapel/blob/main/runtime/src/gpu/amd/gpu-amd.c#L69 I would certainly hope that `hipSetDevice` is relatively no-op-y for the scenario you ask about.

Reply:
Okay from some [references](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html), it looks like it should have minimal overhead for duplicate calls. To my brain it initially seemed that trying to switch contexts a lot might be super costly. Thanks for checking!